docs/design/coreclr/botr/corelib.md

```diff
@@ -0,0 +1,352 @@
+
+### FCall epilog walker for x86
+
+The managed stack walker needs to be able to find its way from FCalls. It is relative easy on newer platforms that define conventions for stack unwinding as part of the ABI. The stack unwinding conventions are not defined by an ABI for x86. The runtime works around this by implementing an epilog walker. The epilog walker computes the FCall return address and callee save registers by simulating the FCall execution. This imposes limits on what constructs are allowed in the FCall implementation.
+
+Complex constructs like stack allocated objects with destructors or exception handling in the FCall implementation may confuse the epilog walker. This can lead to GC holes or crashes during stack walking. There is no comprehensive list of what constructs should be avoided to prevent this class of bugs. An FCall implementation that is fine one day may break with the next C++ compiler update. We depend on stress runs and code coverage to find bugs in this area.
+
+Setting a breakpoint inside an FCall implementation may confuse the epilog walker. It leads to an "Invalid breakpoint in a helpermethod frame epilog" assert inside [vm\i386\gmsx86.cpp](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/i386/gmsx86.cpp).
+
+### FCall example – managed
+
+Here's a real-world example from the `String` class:
+
+```CSharp
+public partial sealed class String
+{        
+    [MethodImpl(MethodImplOptions.InternalCall)]
+    private extern string? IsInterned();
+
+    public static string? IsInterned(string str)
+    {
+        if (str == null)
+        {
+            throw new ArgumentNullException(nameof(str));
+        }
+
+        return str.IsInterned();
+    }
+}
+```
+
+### FCall example – unmanaged
+
+The FCall entrypoint has to be registered in tables in [vm\ecalllist.h][ecalllist] using `FCFuncEntry` macro. See ["Registering your QCall or FCall Method"](#register).
+
+This method is an instance method in managed code, with the "this" parameter passed as the first argument. We use `StringObject*` as the argument type, then copy it into a `STRINGREF` so we get some error checking when we use it.
+
+```C++
+FCIMPL1(Object*, AppDomainNative::IsStringInterned, StringObject* pStringUNSAFE)
+{
+    FCALL_CONTRACT;
+
+    STRINGREF       refString   = ObjectToSTRINGREF(pStringUNSAFE);
+    STRINGREF*      prefRetVal  = NULL;
+
+    HELPER_METHOD_FRAME_BEGIN_RET_1(refString);
+
+    if (refString == NULL)
+        COMPlusThrow(kArgumentNullException, W("ArgumentNull_String"));
+
+    prefRetVal = GetAppDomain()->IsStringInterned(&refString);
+
+    HELPER_METHOD_FRAME_END();
+
+    if (prefRetVal == NULL)
+        return NULL;
+
+    return OBJECTREFToObject(*prefRetVal);
+}
+FCIMPLEND
+```
+
+## <a name="register"></a> Registering your QCall or FCall method
```

Comment:
```suggestion
## <a name="register" /> Registering your QCall or FCall method
```

Reply:
I have hit instances where that doesn't work in markdown. Will update to make it an empty tag.